atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCells.java

```diff
@@ -156,54 +157,65 @@ SweepBatch getBatchForPartition(ShardAndStrategy shardStrategy, long partitionFi
+    private TimestampsToSweep getTimestampsToSweepDescendingAndCleanupAborted(ShardAndStrategy shardStrategy,
+            long minTsExclusive, long sweepTs, Multimap<Long, WriteInfo> writesByStartTs) {
+        Map<Long, Long> startToCommitTs = commitTsCache.loadBatch(writesByStartTs.keySet());
         Map<TableReference, Multimap<Cell, Long>> cellsToDelete = new HashMap<>();
         List<Long> committedTimestamps = new ArrayList<>();
-
-        for (Map.Entry<Long, Long> entry : startToCommitTs.entrySet()) {
-            if (entry.getValue() == TransactionConstants.FAILED_COMMIT_TS) {
-                startTsWrites.get(entry.getKey())
+        long lastSweptTs = minTsExclusive;
+        boolean processedAll = true;
+
+        List<Long> sortedStartTimestamps = startToCommitTs.keySet().stream().sorted().collect(Collectors.toList());
+        for (long startTs : sortedStartTimestamps) {
+            long commitTs = startToCommitTs.get(startTs);
+            if (commitTs == TransactionConstants.FAILED_COMMIT_TS) {
+                lastSweptTs = startTs;
+                writesByStartTs.get(startTs)
                         .forEach(write -> cellsToDelete
                                 .computeIfAbsent(write.tableRef(), ignore -> HashMultimap.create())
                                 .put(write.cell(), write.timestamp()));
+            } else if (commitTs < sweepTs) {
+                lastSweptTs = startTs;
+                committedTimestamps.add(startTs);
             } else {
-                committedTimestamps.add(entry.getKey());
+                processedAll = false;
+                break;
             }
         }
+
         cellsToDelete.forEach((tableRef, multimap) -> {
             kvs.delete(tableRef, multimap);
             maybeMetrics.ifPresent(metrics -> metrics.updateAbortedWritesDeleted(shardStrategy, multimap.size()));
             log.info("Deleted {} aborted writes from the KVS.", SafeArg.of("number", multimap.size()));
         });
 
         committedTimestamps.sort(Comparator.reverseOrder());
-        return committedTimestamps;
+        return TimestampsToSweep.of(committedTimestamps, lastSweptTs, processedAll);
+    }
+
+    private Collection<WriteInfo> getWritesToSweep(Multimap<Long, WriteInfo> writesByStartTs, List<Long> startTs) {
+        Map<CellReference, WriteInfo> writesToSweepFor = new HashMap<>();
+        startTs.stream()
+                .map(writesByStartTs::get)
+                .flatMap(Collection::stream)
+                .forEach(write -> writesToSweepFor
+                        .putIfAbsent(write.writeRef().cellReference(), write));
+        return writesToSweepFor.values();
+    }
+
+    private long getLastSweptTs(TimestampsToSweep startTsCommitted, RowColumnRangeIterator resultIterator,
+            long partitionFine, long maxTsExclusive) {
+        if (startTsCommitted.processedAll() && exhaustedAllColumns(resultIterator)) {
```

Comment:
Remind me when we wouldn't have exhausted all columns?

Reply:
We get the whole row, but we only scan the columns until we go over the threshold of ``SweepQueueUtils.SWEEP_BATCH_SIZE`` cells.